microprofile/WFLY-12590_Integrate_MicroProfile_Fault_Tolerance_2.0.adoc

```diff
@@ -0,0 +1,103 @@
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+
+* The specification repository resides at official Eclipse organization on GitHub in https://github.com/eclipse/microprofile-fault-tolerance[eclipse/microprofile-fault-tolerance] repository.
+* The implementation of Eclipse MicroProfile Fault Tolerance will be provided by SmallRye project https://github.com/smallrye/smallrye-fault-tolerance[smallrye/smallrye-fault-tolerance].
+* The underlying library used by _current_ SmallRye implementation is Hystrix which resides in official Netflix GitHub organization https://github.com/Netflix/Hystrix[Netflix/Hystrix].
+
+=== Other Interested Projects
+
+Not interested in our integration per se, but projects that have already integrated MP FT:
+
+* https://thorntail.io/[Thorntail] has integrated MicroProfile Fault Tolerance as a https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-fault-tolerance[fraction].
+* https://quarkus.io/[Quarkus] has integrated MicroProfile Fault Tolerance as an https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-fault-tolerance[extension].
+
+== Requirements
+
+* Support Eclipse MicroProfile Fault Tolerance version 2.0 for all applications deployed in WildFly.
+* The extension will be provided as Galleon layer called `microprofile-fault-tolerance`.
+* [WILL BE CLARIFIED] The extension will be added to relevant default configurations provided by WildFly (standalone and domain mode).
+* Upstream Eclipse MicroProfile Fault Tolerance TCK test suite will be added to `microprofile-tck` WildFly upstream test suite.
+* Add integration tests above the scope of TCK (e.g. from https://github.com/thorntail/thorntail/tree/master/testsuite/testsuite-microprofile-fault-tolerance[Thorntail])
+
+== Nice-to-Have Requirements
+
+* Create a MicroProfile Fault Tolerance https://github.com/wildfly/quickstart[quickstart] – delivered via separate RFE above of scope of this one.
+
+Items that are currently only being taken into consideration with next version of this feature:
+
+* The underlying library used by SmallRye FT implementation is Hystrix which is currently in "maintenance" mode and a final
+  release of the project (version 1.5.18) has been released. The discussion about replacing the library is ongoing.
+* WildFly's standard management model will be configuration-less
+** all configuration is done through MicroProfile Config specification
+** given the maintenance problem above it is not desirable to expose a management model which will change in near future as concepts are not cleanly mappable to a different underlying library
+* Support for plurality of isolated deployments – not viable with the current implementation, will be possible once smallrye is based on different library than Hystrix.
```

Comment:
@rhusar Here or in the implementation plan section please describe what will happen in this first version of the feature if:
1) User deploys > 1 deployment.
2) User deploys an ear.

Reply:
Will do once I figure it out.

Reply:
@rhusar could you update this section, please? We need to define th behaviour for the two scenarios @bstansberry has mentioned in [1].

[1] https://github.com/wildfly/wildfly-proposals/pull/250#discussion_r342770835

Reply:
@maeste hi, as @rhusar is OOO, could you take a look at above point [1] and provide your pov, please? We need some decision to finish this analyses and test plan.

I had discussion on chat with @rhusar about this topic and major limitation was that hystrix is singleton. So if 2 MP FT deployments are present then they must use the same instance of hystrix. So 2nd deployment cannot apply different configuration for hystrix and must use instance configured by 1st deployment.

We've discussed following options:
1. * to have only single deployment and provide multideployment later (with new MP FT implementation)
2. * to have multi-deployment but any config from microprofile-config.properties would not be applied (just what is set system/env variables and MP config subsystem)
3. * to change hystrix not to be singleton (not sure if it's possible)

There might be other options which we did not discuss. 

I've nacked option 2) as configuration using microprofile-config.properties is must have. 

Which way would you recommend to proceed?

[1] https://github.com/wildfly/wildfly-proposals/pull/250#discussion_r342770835

Reply:
If you look at the current wildfly repo PR, its allowing first deployment to configure and issue the warning on subsequent deployments.

Reply:
3) would require maintaining our own fork and fixes to SR. Thus its 1).

Reply:
thanks @rhusar so in case there are 2 MP FT deployments and server is started then it's race, right?

Reply:
Moved to requirements describing expected behavior.